### PR TITLE
Autoload for mdwn files

### PR DESCRIPTION
**Pull request by sten0**
## Description

This PR Adds support for autoloading markdown-mode when a foo.mdwn file is visited.  It also includes a minor optimisation and cleanup.  Please let me know if you'd like me to rebase and drop any of these commits.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the **CONTRIBUTING.md** document.
- [x] I have updated the documentation in the **README.md** file if necessary.
- [x] I have added an entry to **CHANGES.md**.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed (using `make test`).

### CODE REVIEW

File: markdown-mode.el [hunk 1]
```diff
@@ -30,7 +30,6 @@
 
 ;; See the README.md file for details.
 
-
```
Comment (anarcat) [inline markdown-mode.el]: why did you remove the form feeds? it's actually quite common to have those in elisp modules. i never quite figured out *why*, that said, but i don't see how this relates to the rest of the (otherwise good) changes in this PR.

Comment (sten0) [reply markdown-mode.el]: Ignorance and maybe OCD ;-)  I thought it was noise.  Since then I've read up on them.  Apparently they're sometimes (rarely) used as a kind of section marker.  I've rebased and dropped that commit.

**Comment by sten0** [comment]
@jrblevin there's also a decent discussion on all the new extensions other applications are using for markdown here: https://bugs.freedesktop.org/show_bug.cgi?id=27441

I'm not sure which ones are the most popular ones are for Markdown 1.0-compatible files, but it might be worth documenting somewhere, in the case where adding a large chunk of them to the markdown-mode's autoload alist wouldn't be a good idea.

**Comment by syohex** [comment]
I suppose it is better to use regexp instead of appending list as below

```lisp
(add-to-list 'auto-mode-alist '("\\.\\(?:markdown\\|md\\|mdwn\\)\\'" . markdown-mode))
```

File: markdown-mode.el [hunk 2]
```diff
@@ -9613,9 +9613,11 @@ rows and columns and the column alignment."
   (add-hook 'kill-buffer-hook #'markdown-live-preview-remove-on-kill t t))
 
 ;;;###autoload
-(add-to-list 'auto-mode-alist '("\\.markdown\\'" . markdown-mode))
-;;;###autoload
-(add-to-list 'auto-mode-alist '("\\.md\\'" . markdown-mode))
+(setq auto-mode-alist
```
Comment (anarcat) [inline markdown-mode.el:9616]: this looks strange - why do we reset the entire auto-mode-alist?

i prefered the previous way of doing multiple add-to-list (as in the readme) or using a regex

Comment (rnkn) [reply markdown-mode.el:9616]: Oh yes, very important *not* to set `auto-mode-alist` here, you'll break literally everyone's Emacses.

Comment (anarcat) [reply markdown-mode.el:9616]: well in this case we `setq` with the return value of `append` so it might just work. but yes, bad practice, i'd say.

Comment (rnkn) [reply markdown-mode.el:9616]: I see. Although this will add duplicates to `auto-mode-alist` whenever the file is evaluated. Defo better to use `add-to-list` (because it checks for duplicates first).

**Comment by anarcat** [comment]
superseded by #443

**Comment by syohex** [comment]
I have merged #443

**Comment by sten0** [comment]
Cool, thank you!  Also, I agree the solution that was merged is better in every way :-)
